### PR TITLE
buildbot and bors integration

### PR DESCRIPTION
**Pull request by djichthys**
Testing  buildbot integration

### CODE REVIEW

**Comment by drpdr** [comment]
bors try

File: .buildbot.sh [hunk 1]
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
+# vim: expandtab sts=0 sw=4 smarttab
+set -e
+
+echo "Test buildbot" 
+
```
Comment (drpdr) [inline .buildbot.sh:6]: Try to do a simple `clang-format` check so we are sure that the communication with `unittest` is right.

Comment (ltratt) [reply .buildbot.sh:6]: I don't think we can do clang-format on Boehm: we need to stay in sync with upstream.

However, @djichthys I think we should at least compile Boehm in buildbot, even if we don't run any tests (although running tests would be good too!).

Comment (drpdr) [reply .buildbot.sh:6]: That's right. I had completely forgotten about upstream.

**Comment by djichthys** [comment]
bors try

**Comment by djichthys** [comment]
bors try

**Comment by djichthys** [comment]
Ready for review

File: .buildbot.sh [hunk 2]
```diff
@@ -2,5 +2,31 @@
 # vim: expandtab sts=0 sw=4 smarttab
 set -e
 
-echo "Test buildbot" 
+ROOT_DIR=$(pwd)/..
 
+# arg-1 : Root directory
+# arg-2 : cmake toolchain file
+function build_boehmGC()
```
Comment (ltratt) [inline .buildbot.sh:9]: Bikeshedding but... should this be `build_bdwgc`?

File: .buildbot.sh [hunk 3]
```diff
@@ -2,5 +2,31 @@
...
+{
+    ARCH=$(echo ${2} | cut -d '.' -f 1)
+    SRC_DIR=${1}/build # build/source directory name used for all repos
```
Comment (ltratt) [inline .buildbot.sh:12]: Isn't this literally *not* the source directory -- it's the build directory?

Comment (djichthys) [reply .buildbot.sh:12]: This is where the source is placed by buildbot for all the repositories.

File: .buildbot.sh [hunk 4]
```diff
@@ -2,5 +2,31 @@
 # vim: expandtab sts=0 sw=4 smarttab
 set -e
 
-echo "Test buildbot" 
+ROOT_DIR=$(pwd)/..
```
Comment (ltratt) [inline .buildbot.sh:5]: Hmm, so I find this bit confusing. I *think* that `pwd` is the root of this repository, so why is the `ROOT_DIR` set to its parent?

Comment (ltratt) [reply .buildbot.sh:5]: Looking below, I *think* I now see why you have `SRC_DIR=$1/build`. I think `ROOT_DIR` should not exist as a variable at all and... [I'll add extra comments below...]

File: .buildbot.sh (hunk 3 above)
Comment (ltratt) [reply .buildbot.sh:12]: I think we can remove `SRC_DIR`.

File: .buildbot.sh [hunk 5]
```diff
@@ -2,5 +2,31 @@
...
+    BUILD_OPTS="-DCMAKE_BUILD_TYPE=Debug \
+                -DCMAKE_INSTALL_PREFIX=${INSTALL_DIR} \
+                -Denable_gcj_support=OFF \
+                -Denable_parallel_mark=OFF \
+                -Denable_threads=OFF \
+                -Denable_dynamic_loading=OFF"
+
+
+    mkdir -p ${BUILD_DIR}
+    cmake -B ${BUILD_DIR} -S ${SRC_DIR} ${BUILD_OPTS} -DCMAKE_TOOLCHAIN_FILE=${2}
```
Comment (ltratt) [inline .buildbot.sh:24]: I think this should be:

```
mkdir -p build
cmake -B build -S . ${BUILD_OPTS} -DCMAKE_TOOLCHAIN_FILE=${2}
```

This means that we don't end up writing files *above* the repository's directory, which seems safer (we don't know what the parent directory above this repository contains AFAIK -- but I might be wrong!).

Comment (djichthys) [reply .buildbot.sh:24]: I agree that we should not normally be writing to parent directories at all.  However, CMake discourages in-source-tree building generally.

Comment (ltratt) [reply .buildbot.sh:24]: It's common with `cmake` to write to a non-git-tracked directory *inside* the repository though -- you don't have to write to a parent directory.

**Comment by djichthys** [comment]
Ready for review again

File: .buildbot.sh [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    cmake --build ${BUILD_DIR}
+}
+
+
+build_bdwgc $(pwd) riscv64-purecap.cmake
+
+# Clean up build directories
+rm -rf $(pwd)/bdwgc_build $(pwd)/bdwgc_install
```
Comment (ltratt) [inline .buildbot.sh:30]: We don't need to clean-up after ourselves (provided -- as you have! -- you've written into the directory that `.buildbot.sh` started execution in), so we can kill this line.

File: .buildbot.sh [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+                -Denable_dynamic_loading=OFF"
+
+
+    mkdir -p ${BUILD_DIR}
+    cmake -B ${BUILD_DIR} -S ${SRC_DIR} ${BUILD_OPTS} -DCMAKE_TOOLCHAIN_FILE=${2}
+    cmake --build ${BUILD_DIR}
+}
+
+
+build_bdwgc $(pwd) riscv64-purecap.cmake
```
Comment (ltratt) [inline .buildbot.sh:27]: Do we need the first argument? `$(pwd)` is equivalent to `.` here, so I think we can kill this argument, remove the `SRC_DIR` argument above, and then simplify things like `BUILD_DIR=${SRC_DIR}/bdwgc_build` away too (it's equivalent to `bdwgc_build`).

**Comment by djichthys** [comment]
Ready for review again

**Comment by ltratt** [comment]
OK this is looking good for a first stage!

**Comment by ltratt** [comment]
bors try

**Comment by ltratt** [comment]
Please squash.

**Comment by djichthys** [comment]
Ready to merge @ltratt .

File: .buildbot.sh [hunk 8]
```diff
@@ -0,0 +1,23 @@
+#!/bin/bash
```
Comment (ltratt) [inline .buildbot.sh:1]: Sorry, I just noticed these two. Unless we really really really need `bash`, we should do `#! /bin/sh` here. Also we shouldn't include your vim settings on the line below. Sorry for not spotting these earlier!

Comment (djichthys) [reply .buildbot.sh:1]: Currently /bin/sh is linked to `dash`  @ltratt .   Should the rest of the script also be changed to work with dash ? or should be change the default shell in the container to `/bin/bash` ?

Comment (ltratt) [reply .buildbot.sh:1]: Ideally the script should work with any `sh` descendent and shouldn't rely on bash or dash (or zsh or ...): I doubt there's anything in here which cares if it's bash or dash.

Comment (djichthys) [reply .buildbot.sh:1]: Regarding the vim settings. This is what I get for copy pasting code from other repos. Will remove!

**Comment by djichthys** [comment]
ready for review again @ltratt

File: .buildbot.sh [hunk 9]
```diff
@@ -1,6 +1,4 @@
-#!/bin/bash
-# vim: expandtab sts=0 sw=4 smarttab
-set -e
```
Comment (ltratt) [inline .buildbot.sh]: We want to keep `set -e` -- it's a wonderful flag!

**Comment by djichthys** [comment]
Ready for review again @ltratt

**Comment by ltratt** [comment]
Please squash.

**Comment by djichthys** [comment]
Ready to merge @ltratt

**Comment by ltratt** [comment]
I didn't notice that this PR wasn't set to merge into master :(

>  merged 1 commit into capablevms:riscv/develop from djichthys:riscv/dj/cicd-setup

I think you might need to raise a new PR with the same files as this one, but that wants to merge into master.

**Pull request merged**